flink/src/main/java/org/apache/iceberg/flink/sink/RowDataTaskWriter.java

```diff
@@ -0,0 +1,141 @@
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.PartitionKey;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.flink.RowDataWrapper;
+import org.apache.iceberg.io.BaseTaskWriter;
+import org.apache.iceberg.io.FileAppenderFactory;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.OutputFileFactory;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.Tasks;
+
+public class RowDataTaskWriter extends BaseTaskWriter<RowData> {
+
+  private static final PartitionKey UNPARTITIONED_KEY = new PartitionKey(PartitionSpec.unpartitioned(), null);
+
+  private final Schema schema;
+  private final PartitionKey partitionKey;
+  private final RowDataWrapper rowDataWrapper;
+  private final List<Integer> equalityFieldIds;
+
+  private final Map<PartitionKey, RowDataDeltaWriter> deltaWriterMap = Maps.newHashMap();
+
+  RowDataTaskWriter(PartitionSpec spec,
+                    FileFormat format,
+                    FileAppenderFactory<RowData> appenderFactory,
+                    OutputFileFactory fileFactory,
+                    FileIO io,
+                    long targetFileSize,
+                    Schema schema,
+                    RowType flinkSchema,
+                    List<Integer> equalityFieldIds) {
+    super(spec, format, appenderFactory, fileFactory, io, targetFileSize);
+
+    this.schema = schema;
+    this.equalityFieldIds = equalityFieldIds;
+
+    this.partitionKey = new PartitionKey(spec, schema);
+    this.rowDataWrapper = new RowDataWrapper(flinkSchema, schema.asStruct());
+  }
+
+  @Override
+  public void write(RowData row) throws IOException {
+    RowDataDeltaWriter deltaWriter;
+
+    if (spec().fields().size() <= 0) {
+      // Create and cache the delta writer if absent for unpartitioned table.
+      deltaWriter = deltaWriterMap.get(UNPARTITIONED_KEY);
+      if (deltaWriter == null) {
+        deltaWriter = new RowDataDeltaWriter(null, equalityFieldIds);
+        deltaWriterMap.put(UNPARTITIONED_KEY, deltaWriter);
+      }
+    } else {
+      // Refresh the current partition key.
+      partitionKey.partition(rowDataWrapper.wrap(row));
+
+      // Create and cache the delta writer if absent for partitioned table.
+      deltaWriter = deltaWriterMap.get(partitionKey);
+      if (deltaWriter == null) {
+        // NOTICE: we need to copy a new partition key here, in case of messing up the keys in writers.
+        PartitionKey copiedKey = partitionKey.copy();
+        deltaWriter = new RowDataDeltaWriter(copiedKey, equalityFieldIds);
+        deltaWriterMap.put(copiedKey, deltaWriter);
+      }
+    }
+
+    switch (row.getRowKind()) {
+      case INSERT:
+      case UPDATE_AFTER:
+        deltaWriter.write(row);
+        break;
+
+      case DELETE:
+      case UPDATE_BEFORE:
+        deltaWriter.delete(row);
+        break;
```

Comment:
I'm not familiar with `UPDATE_AFTER` or `UPDATE_BEFORE`. Can you help me understand what's going on here?

Reply:
For CDC events,  such as mysql binlog: `UPDATE test set a=2, b=2 where a=1, b=1`,  the [flink-cdc-connectors](https://github.com/ververica/flink-cdc-connectors) will produce two change logs for it:

```
-U (a=1, b=1)
+U (a=2,b=2)
```

The first `RowData` means we will need to delete the old row (1,1),  it's also called a `UPDATE_BEFORE`  row. The second `RowData` means we will need to insert the new row (2,2),  it's also called a `UPDATE_AFTER` row.

Reply:
What distinguishes `UPDATE_BEFORE` from `DELETE`? Is it that it will be associated with `UPDATE_AFTER` so the two are combined to form a replace operation?

Reply:
> Is it that it will be associated with UPDATE_AFTER so the two are combined to form a replace operation?

Yes, that's right.